sql: implement the configuration variable `bytea_output`

knz opened the pull request:
Fixes  #20782.
Fixes #25841.

PostgreSQL specifies that this session variable determines how byte
arrays are converted to strings, at least in the following contexts:

- when byte arrays are communicated back to the client, in the text
  protocol,
- when byte array values are casted to a string type in a scalar
  expression.
- in the `xxx_output` internal functions (not yet implemented by
  CockroachDB).

There are two values possible as per PostgreSQL, either `hex` or
`escape`. Incidentally, these encoding is identical as that supported
by the `encode()` and `decode()` built-ins.

While the `hex` format is trivial, the `escape` format is
"interesting": it is much more restricted than the
"string-with-escapes" format recognized by CockroachDB's input
scanner. In particular, PostgreSQL only allows the special sequences
`\\` (to specify a single `\` char) and `\NNN` with 3 octal
digits. Hex escapes and other C-like single-char escapes are not
supported.

This matters because this implies that *clients are not expecting to
see non-octal, non-\\ escape escape sequences* in byte arrays rendered
with `escape`. The previous logic used by CockroachDB to encode byte
arrays as strings is being restricted in this patch to make this true.

To achieve this change, this patch is organized as follows:

- two new functions `DecodeRawBytesToByteArray` and
  `EncodeByteArrayToRawBytes` are created in package `lex`, able to
  recognize the 3 known encodings (`hex`, `escape` and `base64`) via
  a parameter.
- a new function `DecodeRawBytesToByteArrayAuto` able to auto-detect
  the format to use when decoding, based on the presence of a `\x`
  prefix.
- the new functions are now reused in multiple places:
  - the `encode`/`decode` built-in functions;
  - the `string` <-> `bytes` casts;
  - the communication of byte arrays on pgwire;
  - the implicit conversion from string literals (`StrVal`)
    to actual SQL values (`Datum`);
  - the parsing of "naked" strings in memory via `tree.ParseDBytes()`.

This patch also removes the auto-conversion from the byte array
literal syntax to INET, which was never tested nor documtned, and was
implemented in a non-intuitive way: one would expect that the raw
bytes of the byte literal would be used to encode the IP address,
whereas the code re-parsed the bytes as decimal/hex digits. This patch
removes that auto-typing rule to avoid any confusion. It may be added
back with the proper semantics in the future.

Finally, this patch also removes the "dynamic typing rule" from the
byte array literal syntax to either the SQL types `string` or
`bytes`. Prior to this patch, the set of candidate types available for
a byte array literal constant was determined depending on whether the
actual values of the bytes formed a valid UTF-8
sequence (incidentally, thereby incurring the cost of a UTF-8
validation in all cases). This would cause surprising behavior where
the same query could type differently depending on the contents of the
string literal. This path is now removed (and the extra UTF-8 scan
cost averted).

Note that the client protocol change triggered by `bytea_output` is
not directly visible in `cockroach sql`: the driver in that
client (`lib/pq`) automatically parses the `hex` format and convert it
back to a byte array client-side which was then pretty-printed without
hex encoding, so byte arrays appear as if `byte_output = escape` was
used in all cases. To observe the difference, use `psql` instead.

Release note (sql change): the session variable `bytea_output` now
controls how byte arrays are converted to strings and reported back to
clients, for compatibility with PostgreSQL.

knz committed:
sql: implement the configuration variable `bytea_output` + bytea cleanup

PostgreSQL specifies that this session variable determines how byte
arrays are converted to strings, at least in the following contexts:

- when byte arrays are communicated back to the client, in the text
  protocol,
- when byte array values are casted to a string type in a scalar
  expression.
- in the `xxx_output` internal functions (not yet implemented by
  CockroachDB).

There are two values possible as per PostgreSQL, either `hex` or
`escape`. Incidentally, these encoding is identical as that supported
by the `encode()` and `decode()` built-ins.

While the `hex` format is trivial, the `escape` format is
"interesting": it is much more restricted than the
"string-with-escapes" format recognized by CockroachDB's input
scanner. In particular, PostgreSQL only allows the special sequences
`\\` (to specify a single `\` char) and `\NNN` with 3 octal
digits. Hex escapes and other C-like single-char escapes are not
supported.

This matters because this implies that *clients are not expecting to
see non-octal, non-\\ escape escape sequences* in byte arrays rendered
with `escape`. The previous logic used by CockroachDB to encode byte
arrays as strings is being restricted in this patch to make this true.

To achieve this change, this patch is organized as follows:

- two new functions `DecodeRawBytesToByteArray` and
  `EncodeByteArrayToRawBytes` are created in package `lex`, able to
  recognize the 3 known encodings (`hex`, `escape` and `base64`) via
  a parameter.
- a new function `DecodeRawBytesToByteArrayAuto` able to auto-detect
  the format to use when decoding, based on the presence of a `\x`
  prefix.
- the new functions are now reused in multiple places:
  - the `encode`/`decode` built-in functions;
  - the `string` <-> `bytes` casts;
  - the communication of byte arrays on pgwire;
  - the implicit conversion from string literals (`StrVal`)
    to actual SQL values (`Datum`);
  - the parsing of "naked" strings in memory via `tree.ParseDBytes()`.

This patch also removes the auto-conversion from the byte array
literal syntax to INET, which was never tested nor documtned, and was
implemented in a non-intuitive way: one would expect that the raw
bytes of the byte literal would be used to encode the IP address,
whereas the code re-parsed the bytes as decimal/hex digits. This patch
removes that auto-typing rule to avoid any confusion. It may be added
back with the proper semantics in the future.

Finally, this patch also removes the "dynamic typing rule" from the
byte array literal syntax to either the SQL types `string` or
`bytes`. Prior to this patch, the set of candidate types available for
a byte array literal constant was determined depending on whether the
actual values of the bytes formed a valid UTF-8
sequence (incidentally, thereby incurring the cost of a UTF-8
validation in all cases). This would cause surprising behavior where
the same query could type differently depending on the contents of the
string literal. This path is now removed (and the extra UTF-8 scan
cost averted).

Note that the client protocol change triggered by `bytea_output` is
not directly visible in `cockroach sql`: the driver in that
client (`lib/pq`) automatically parses the `hex` format and convert it
back to a byte array client-side which was then pretty-printed without
hex encoding, so byte arrays appear as if `byte_output = escape` was
used in all cases. To observe the difference, use `psql` instead.

Release note (sql change): the session variable `bytea_output` now
controls how byte arrays are converted to strings and reported back to
clients, for compatibility with PostgreSQL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25835)
<!-- Reviewable:end -->

knz commented:
@justinj can you have a preliminary look? I know this PR is currently missing tests, but before I go the testing route I'd like to know if the approach I've taken is correct.

knz committed:
sql/parser: clean up the scanner

Once upon the time the scanner allowed one to customize the identifier
and string delimiter characters. This is not the case any more. This
patch cleans up the last remnants of this artifact.

Also, the code to scan hexadecimal byte arrays was very inefficient as
it would require multiple successive copies of the data. This patch
specializes the code to avoid the extraneous copies.

Release note: None

knz commented:
Now with tests, ready to review. Thanks!

knz commented:
@dt can we please sit together next to this code. The IMPORT code for MySQL data was relying on an invalid behavior of the byte array parsing code, which needs to go away for compatibility. I want to understand what needs to be modified so that the IMPORT code can use its own decoding routines. Thanks.

knz commented:
Symmary of the discussion with @dt and @mjibson:

1) spec: the column formats recognized by regular IMPORT CSV are exactly those we allow inside simple (unescaped) string literals in SQL. i.e. if the value `'XXXX'` is valid when cast to type T, then the value `XXXX` (without the simple quotes) is valid in a CSV column. 

2) Both the code for IMPORT CSV and what we have documented already is properly aligned with this spec.

3) postgres specifies that when converting a SQL string to BYTES/BYTEA, only the following quoting styles are recognized: `\\` for `\`, and `\NNN` for a byte with octal encoding NNN.

4) in particular, if a SQL string `'XXXX'` contains garbage/random bytes including `\` followed by neither `\` nor 3 octal digits, then the conversion of that string to BYTES/BYTEA will fail, as per pg's specs (and enforced in this PR)

4) currently a MySQL output file for a byte array will contain garbage/random bytes and possibly unescaped `\` characters. This makes these values unsuitable for processing by IMPORT CSV, and `TestImportMysqlOutput` is thus attempting an invalid conversion.

5) @mjibson is currently addressing point 4 in a separate PR so that importing MySQL data will not use the CSV reader.

So I am disabling `TestImportMysqlOutput` here until Matt's PR makes the code correct.

knz commented:
This PR is thus ready for review. ping @justinj 

and perhaps @mjibson @dt you can also have a look

mjibson commented:
We will have to add docs about the `\\` and `\NNN` escapes to IMPORT CSV (which is a change, but a good one I think), but otherwise this LGTM from a high level (I didn't review all the encoding logic, just the descriptions of what would change here).

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 33 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/lex/encode.go, line 265 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBJbPLDJpYA--kD519:-LEBJbPLDJpYA--kD51A:bvqj40p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode.go#L265)):*
> ```
> https://www.postgresql.org/docs/current/static/datatype-binary.html#AEN5667
> ```
this says:
>The requirement to escape non-printable octets varies depending on locale settings

is this something we need to be concerned about?

---

*[pkg/sql/lex/encode.go, line 315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBK4O78cbVu1k8VXk6:-LEBK4O83eR_2vVQIZM6:b-j2dqgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode.go#L315)):*
> ```Go
> 			b := byte(0)
> 			for j := 1; j <= 3; j++ {
> 				if i+j >= len(data) {
> ```

nit: it might be cleaner (and slightly faster) to do this check outside of this loop

---

*[pkg/sql/lex/encode.go, line 335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBKKvN6YjspiNbR-kq:-LEBKKvN6YjspiNbR-kr:b-fj4y21) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode.go#L335)):*
> ```Go
> 
> 	default:
> 		panic(fmt.Sprintf("unhandled format: %s", be))
> ```

nit: since we can already error out of this method, maybe we might as well do the slightly safer thing and just error instead of panicking?

---

*[pkg/sql/lex/encode_test.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBKwmm6Jkijg8nETqg:-LEBKwmn0xh8t0RvEpMr:bdul37l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode_test.go#L169)):*
> ```Go
> 		{`/w==`, false, fmtB64, "\xff", ""},
> 		{`AAAA`, false, fmtB64, "\x00\x00\x00", ""},
> 		{`a`, true, 0, "a", ""},
> ```

I'd be tempted to make the auto and non-auto two different tests, but doesn't matter too much either way

---

*[pkg/sql/parser/scan.go, line 833 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBIev0-mVD1FUEu9ew:-LEBIev0-mVD1FUEu9ex:b9m6vjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1808743636841071a0d9755213a2c9e6b8c9f899/pkg/sql/parser/scan.go#L833)):*
> ```Go
> }
> 
> func (s *Scanner) scanString(lval *sqlSymType, ch int, allowEscapes, requireUTF8 bool) bool {
> ```

I think this function would benefit from a comment that explains what kinds of strings (like, which formats) are allowed here.

---

*[pkg/sql/pgwire/types_test.go, line 154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBM54VCFbCi0Q2MbCL:-LEBM54VCFbCi0Q2MbCM:b-r6u2nq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/pgwire/types_test.go#L154)):*
> ```Go
> 			buf := newWriteBuffer(nil /* bytecount */)
> 			buf.bytecount = metric.NewCounter(metric.Metadata{})
> 			d := tree.NewDBytes(tree.DBytes("\x00abc\\\n"))
> ```

maybe a good place for a use of `RandDatum` if possible? or some kind of generation?

---

*[pkg/sql/sem/tree/constant.go, line 352 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBMRZB0QGF45O9iTO1:-LEBMRZB0QGF45O9iTO2:b-hv7ncf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/sem/tree/constant.go#L352)):*
> ```Go
> 	// x'....'. If false, the string is guaranteed to be a valid UTF-8
> 	// sequence.
> 	scannedAsBytes bool
> ```

yessss

---

*[pkg/sql/sem/tree/constant.go, line 465 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBMYo-BrzjKZi3KZLx:-LEBMYo04AQOdM21zsnF:bry3y3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/sem/tree/constant.go#L465)):*
> ```Go
> 			expr.resString = DString(expr.s)
> 			return &expr.resString, nil
> 		case types.INet:
> ```

what's the reason for this clause?

---

*[pkg/sql/sem/tree/constant.go, line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBMa5TDOkm762DLpla:-LEBMa5TDOkm762DLplb:blqgxzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/sem/tree/constant.go#L588)):*
> ```Go
> 				case Concat:
> 					// When folding string-like constants, if either was a byte
> 					// array ltieral, the result is also a byte literal.
> ```

nit: `literal`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-:-LEBMnEL05-qVVQklIkY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 33 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lex/encode.go, line 265 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBJbPLDJpYA--kD519:-LEEFkK6-4hTQyN4g3W5:bfjjg0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode.go#L265)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

> ```
> https://www.postgresql.org/docs/current/static/datatype-binary.html#AEN5667
> ```
this says:
>The requirement to escape non-printable octets varies depending on locale settings

is this something we need to be concerned about?
</blockquote></details>

Added a note in a comment.

---

*[pkg/sql/lex/encode.go, line 315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBK4O78cbVu1k8VXk6:-LEEG7oB3-LAMv5QNk8w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode.go#L315)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: it might be cleaner (and slightly faster) to do this check outside of this loop
</blockquote></details>

Done.

---

*[pkg/sql/lex/encode.go, line 335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBKKvN6YjspiNbR-kq:-LEEGFQv4r63CozngUUF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode.go#L335)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: since we can already error out of this method, maybe we might as well do the slightly safer thing and just error instead of panicking?
</blockquote></details>

Done.

---

*[pkg/sql/lex/encode_test.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBKwmm6Jkijg8nETqg:-LEEGtpp8Suw5f_b7rGr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/lex/encode_test.go#L169)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I'd be tempted to make the auto and non-auto two different tests, but doesn't matter too much either way
</blockquote></details>

Done.

---

*[pkg/sql/parser/scan.go, line 833 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBIev0-mVD1FUEu9ew:-LEEHC9A6vGpzavOgAV5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1808743636841071a0d9755213a2c9e6b8c9f899/pkg/sql/parser/scan.go#L833)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think this function would benefit from a comment that explains what kinds of strings (like, which formats) are allowed here.
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/types_test.go, line 154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBM54VCFbCi0Q2MbCL:-LEEJjtM6bA6-C3i76Il:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/pgwire/types_test.go#L154)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

maybe a good place for a use of `RandDatum` if possible? or some kind of generation?
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/constant.go, line 465 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBMYo-BrzjKZi3KZLx:-LEEKFrE4QDEoI4SxIbv:bqcjq78) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/sem/tree/constant.go#L465)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

what's the reason for this clause?
</blockquote></details>

It's was an error. Thanks for catching it. Fixed.

---

*[pkg/sql/sem/tree/constant.go, line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25835#-LEBMa5TDOkm762DLpla:-LEEKP7Y4JLx3Hrl55cq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e646d4d0c6f964bba15e4293aa79cd4b26b30a23/pkg/sql/sem/tree/constant.go#L588)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: `literal`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25835)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for the reviews!

bors r+

knz commented:
forgot to change logic test...

bors r+
